### DISCUSSION TITLE
Contributing new models

### DISCUSSION THREAD

**Discussion by thekevinscott** [thekevinscott/UpscalerJS#32] [Ideas]
I'd like to explore making it easier to submit new models to the repo.

To train a new model, a developer needs to be willing to sling some Python. However, I think there's ways we can ease the process.

Google Colab offers free GPUs, and I _believe_ that training idealo's implementation of ESRGAN can fit within the Colab time limits. (Note to self: confirm this.)

Training a brand new architecture from scratch is difficult in that it requires ascertaining that the Python implementation is compatible with Tensorflow.js. However, re-training an _existing_ implementation could be a lot more straight forward - find a dataset, set up the code to split the dataset into appropriate train / test sets (if not already done), run the training on Colab and babysit as needed, and then submit the model to this repo.

Specific datasets ([like Flickr's faces dataset](https://github.com/NVlabs/ffhq-dataset)) could allow us to develop more domain-specific models that offer better accuracy for specific domains.

Some open questions:

- Can Colab support this process? If not, is there a service we could use to train models that is free? (Introducing a paid tier would I think significantly hinder abilities to recruit users willing to train new models)
- How much can we automate the process of integrating a new dataset into the training script?
- What does the submission process look like? I assume we'd want to record information about how the model was trained - even maybe include a copy of the notebook associated with training. Automatic benchmarking would also be nice. Would a model be auto-deployed, or would we need to cut a new release?
